tools/common/FileCopier.cs

```diff
@@ -0,0 +1,249 @@
+			// so we need to use native functions directly. Luckily OSX provides exactly what we need.
+			IntPtr state = copyfile_state_alloc ();
+			try {
+				CopyFileCallbackDelegate del = CopyFileCallback;
+				copyfile_state_set (state, CopyFileState.StatusCB, Marshal.GetFunctionPointerForDelegate (del));
+				int rv = copyfile (source, target, state, CopyFileFlags.Data | CopyFileFlags.Recursive | CopyFileFlags.Nofollow | CopyFileFlags.Clone);
+				if (rv != 0)
+					throw CreateError (1022, "Could not copy the directory '{0}' to '{1}': {2}", source, target, strerror (Marshal.GetLastWin32Error ()));
+			} finally {
+				copyfile_state_free (state);
+			}
+		}
+
+		static CopyFileResult CopyFileCallback (CopyFileWhat what, CopyFileStep stage, IntPtr state, string source, string target, IntPtr ctx)
+		{
+//			Console.WriteLine ("CopyFileCallback ({0}, {1}, 0x{2}, {3}, {4}, 0x{5})", what, stage, state.ToString ("x"), source, target, ctx.ToString ("x"));
+			switch (what) {
+			case CopyFileWhat.File:
+				if (!IsUptodate (source, target)) {
+					if (stage == CopyFileStep.Finish)
+						Log (1, "Copied {0} to {1}", source, target);
+					return CopyFileResult.Continue;
+				} else {
+					Log (3, "Target '{0}' is up-to-date", target);
+					return CopyFileResult.Skip;
+				}
+			case CopyFileWhat.Dir:
+			case CopyFileWhat.DirCleanup:
+			case CopyFileWhat.CopyData:
+			case CopyFileWhat.CopyXattr:
+				return CopyFileResult.Continue;
+			case CopyFileWhat.Error:
+				throw CreateError (1021, "Could not copy the file '{0}' to '{1}': {2}", source, target, strerror (Marshal.GetLastWin32Error ()));
+			default:
+				return CopyFileResult.Continue;
+			}
+		}
+
+		// Checks if the source file has a time stamp later than the target file.
+		//
+		// Optionally check if the contents of the files are different after checking the timestamp.
+		//
+		// If check_stamp is true, the function will use the timestamp of a "target".stamp file
+		// if it's later than the timestamp of the "target" file itself.
+		public static bool IsUptodate (string source, string target, bool check_contents = false, bool check_stamp = true)
+		{
+			if (Force)
+				return false;
+
+			var tfi = new FileInfo (target);
+			
+			if (!tfi.Exists) {
+				Log (3, "Target '{0}' does not exist.", target);
+				return false;
+			}
+
+			if (check_stamp) {
+				var tfi_stamp = new FileInfo (target + ".stamp");
+				if (tfi_stamp.Exists && tfi_stamp.LastWriteTimeUtc > tfi.LastWriteTimeUtc) {
+					Log (3, "Target '{0}' has a stamp file with newer timestamp ({1} > {2}), using the stamp file's timestamp", target, tfi_stamp.LastWriteTimeUtc, tfi.LastWriteTimeUtc);
+					tfi = tfi_stamp;
+				}
+			}
+
+			var sfi = new FileInfo (source);
+
+			if (sfi.LastWriteTimeUtc <= tfi.LastWriteTimeUtc) {
+				Log (3, "Prerequisite '{0}' is older than the target '{1}'.", source, target);
+				return true;
+			}
+
+			if (check_contents && CompareFiles (source, target)) {
```

Comment:
I would do this:

```csharp
#if MMP || MTOUCH
    if (check_contents ...) {
        // ...
    }
#else
    if (check_contents)
        throw new NotImplementedException ("Checking file contents is not supported");
#endif
```

and then you can remove all the Compare* functions completely:

```csharp
#if !MMP && !MTOUCH
    [DllImport ("/usr/lib/libSystem.dylib", SetLastError = true, EntryPoint = "strerror")]
    static extern IntPtr _strerror (int errno);
    internal static string strerror (int errno) {}
    public static bool CompareFiles (string a, string b, bool ignore_cache = false) {}
    public unsafe static bool CompareStreams (Stream astream, Stream bstream, bool ignore_cache = false) {}
#endif
```